File: worker/uri-parser.ts
export interface ServerUri {
    type: 'smp' | 'xftp';
    domain: string;
};

const domainPattern = /(?:[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?\.)+[a-zA-Z]{2,}/g;
const extractDomain = function (inputString: string): string | null {
    const matches = inputString.match(domainPattern);
    return matches ? matches[0] : null;
}

export const parseUri = function (uri: string): ServerUri {
    const type = uri.split(':')[0] as 'smp' | 'xftp';
    if (uri.endsWith('.onion')) {
        if (uri.includes(',')){
            return {
                type,
                domain: extractDomain(uri),
            };
        } else {
            return {
                type,
                domain: null,
            };
        }
    } else {
        return {
            type,
            domain: extractDomain(uri),
        }
    }
};

export const isServerOfficial = function (uri: string): boolean {
    return uri.includes('simplex.im');
};

File: validator/src/uri_parser.rs
use regex::Regex;

fn extract_hostnames<'a>(input: &'a str) -> impl Iterator<Item = &'a str> {
    let host_part = if let Some(value) = input.split('@').nth(1) {
        value
    } else {
        ""
    };
    host_part.split(',')
}

pub fn parse_uri<'a>(uri: &'a str) -> Result<impl Iterator<Item = &'a str>, Box<dyn std::error::Error>> {
    let uri_type = uri.split(':').next().unwrap_or("").to_string();
    if uri_type != "smp" && uri_type != "xftp" {
        return Err("Invalid SMP/XFTP URI".into())
    }

    Ok(extract_hostnames(uri).filter(|hostname| !is_server_official(hostname)))
}

pub fn is_server_official(uri: &str) -> bool {
    uri.contains("simplex.im")
}
